File: Math/233_NumberOfDigitOne.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# @Author: xuezaigds@gmail.com


class Solution(object):
    # Recursive solution
    def countDigitOne(self, n):
        if n <= 0:
            return 0
        elif n < 10:
            return 1
        else:
            units = n % 10
            tens = n / 10
            count = self.countDigitOne(tens - 1) * 10 + tens
            n /= 10
            while n:
                if n % 10 == 1:
                    count = count + 1 + units
                n = n / 10

            if units >= 1:
                count += 1
            return count

"""
-1
6
12
234545
"""

File: Math/233_NumberOfDigitOne.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-04-28 13:58:30
 */

class Solution {
public:
    int countDigitOne(int n)
    {
        if(n<=0) return 0;
        long long count = 1;
        while(n){
            if(n<10){
                break;
            }
            int digit = n % 10;
            n /= 10;
            count += n;
            if(digit == 0)  count -= 1;
            count += countDigitOne(n-1) * 10;

            // 最后一行中数组1出现的次数
            while(n){
                if(n%10==1) count += digit+1;
                n /= 10;
            }
        }
        return count;
    }
};

/*
-1
6
12
234545
*/
